mutex/mutexworkflow.py

```diff
@@ -0,0 +1,117 @@
+from dataclasses import dataclass
+from temporalio.client import Client
+from temporalio import activity, workflow
+import asyncio
+from temporalio.exceptions import ApplicationError
+
+LOCK_ACQUIRED_SIGNAL_NAME = "acquire-lock-event"
+
+
+@dataclass
+class SignalWithStartMutexWorkflowInput:
+    namespace: str
+    resource_id: str
+    unlock_timeout_seconds: float
+    sender_workflow_id: str
+
+
+@dataclass
+class SignalWithStartMutexWorkflowResult:
+    workflow_id: str
+
+
+@dataclass
+class MutexWorkflowInput:
+    namespace: str
+    resource_id: str
+    unlock_timeout_seconds: float
+
+
+@activity.defn
+async def signal_with_start_mutex_workflow(
+    params: SignalWithStartMutexWorkflowInput,
+) -> SignalWithStartMutexWorkflowResult:
+    # Create client connected to server at the given address
+    client = await Client.connect("localhost:7233")
+
+    workflow_id = f"mutex:{params.namespace}:{params.resource_id}"
+
+    # Sends a signal to the workflow (and starts it if needed)
+    wf_input = MutexWorkflowInput(
+        namespace=params.namespace,
+        resource_id=params.resource_id,
+        unlock_timeout_seconds=params.unlock_timeout_seconds,
+    )
+    await client.start_workflow(
+        workflow=MutexWorkflow.run,
+        arg=wf_input,
+        id=workflow_id,
+        task_queue="mutex-task-queue",
+        start_signal="request_lock",
+        start_signal_args=[params.sender_workflow_id],
+    )
+    return SignalWithStartMutexWorkflowResult(workflow_id=workflow_id)
+
+
+def generate_unlock_token(sender_workflow_id: str) -> str:
+    return f"unlock-event-{sender_workflow_id}"
+
+
+@workflow.defn
```

Comment:
I think we may not want to blind-port this from Go but have a more Pythonic way to do a mutex and we want to make sure workflows are in their own files separate from activities/models. For instance, the .NET version of this sample at https://github.com/temporalio/samples-dotnet/tree/main/src/Mutex made an idiomatic `WorkflowMutex` class for general purpose use. I think this sample should consider a workflow-backed `Lock` class that looks almost exactly like https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock and is copyable/usable as if it is a library.